server/data/data-jpa/src/main/java/org/apache/james/sieve/jpa/JPASieveRepository.java

```diff
@@ -0,0 +1,345 @@
+
+    @Override
+    public InputStream getScript(User user, ScriptName name) throws ScriptNotFoundException, StorageException {
+        Optional<JPASieveScript> script = findSieveScript(user, name);
+        JPASieveScript sieveScript = script.orElseThrow(() -> new ScriptNotFoundException("Unable to find script " + name.getValue() + " for user " + user.asString()));
+        return IOUtils.toInputStream(sieveScript.getScriptContent(), StandardCharsets.UTF_8);
+    }
+
+    private Optional<JPASieveScript> findSieveScript(User user, ScriptName scriptName) throws StorageException {
+        return transactionRunner.runAndRetrieveResult(entityManager -> findSieveScript(user, scriptName, entityManager),
+                throwStorageException("Unable to find script " + scriptName.getValue() + " for user " + user.asString()));
+    }
+
+    private Optional<JPASieveScript> findSieveScript(User user, ScriptName scriptName, EntityManager entityManager) {
+        try {
+            JPASieveScript sieveScript = entityManager.createNamedQuery("findSieveScript", JPASieveScript.class)
+                    .setParameter("username", user.asString())
+                    .setParameter("scriptName", scriptName.getValue()).getSingleResult();
+            return Optional.ofNullable(sieveScript);
+        } catch (NoResultException e) {
+            LOGGER.debug("Sieve script not found for user {}", user.asString());
+            return Optional.empty();
+        }
+    }
+
+    @Override
+    public void deleteScript(User user, ScriptName name) throws ScriptNotFoundException, IsActiveException, StorageException {
+        transactionRunner.runAndHandleException(Throwing.<EntityManager>consumer(entityManager -> {
+            Optional<JPASieveScript> sieveScript = findSieveScript(user, name, entityManager);
+            if (!sieveScript.isPresent()) {
+                rollbackTransactionIfActive(entityManager.getTransaction());
+                throw new ScriptNotFoundException("Unable to find script " + name.getValue() + " for user " + user.asString());
+            }
+            JPASieveScript sieveScriptToRemove = sieveScript.get();
+            if (sieveScriptToRemove.isActive()) {
+                rollbackTransactionIfActive(entityManager.getTransaction());
+                throw new IsActiveException("Unable to delete active script " + name.getValue() + " for user " + user.asString());
+            }
+            entityManager.remove(sieveScriptToRemove);
+        }).sneakyThrow(), throwStorageException("Unable to delete script " + name.getValue() + " for user " + user.asString()));
+    }
+
+    @Override
+    public void renameScript(User user, ScriptName oldName, ScriptName newName) throws ScriptNotFoundException, DuplicateException, StorageException {
+        transactionRunner.runAndHandleException(Throwing.<EntityManager>consumer(entityManager -> {
+            Optional<JPASieveScript> sieveScript = findSieveScript(user, oldName, entityManager);
```

Comment:
I usually use ifPresent in simpler constructions. Here we would have ifPresent inside ifPresent. It would like a little bit more messy in my opinion. Beside that I will have to rollback transaction and throw exception when no script found.